### ISSUE TITLE
FP: cast_lossless

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9882]
### Summary

.

### Lint Name

cast_lossless

### Reproducer

I tried this code:

```rust
#![feature(repr128)]

#[derive(PartialEq, Debug)]
#[repr(i128)]
enum Test {
    A(Box<u64>) = 0,
    B(usize) = u64::MAX as i128 + 1,
}

fn main() {
    assert_ne!(Test::A(Box::new(2)), Test::B(0));
}

```

I saw this happen:
`cargo clippy  --fix -- -Aclippy::all -Wclippy::cast_lossless`
```
    Checking a v0.1.0 (/tmp/a)
warning: failed to automatically apply fixes suggested by rustc to crate `a`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
warning: the feature `repr128` is incomplete and may not be safe to use and/or cause compiler crashes
 --> src/main.rs:1:12
  |
1 | #![feature(repr128)]
  |            ^^^^^^^
  |
  = note: see issue #56071 <https://github.com/rust-lang/rust/issues/56071> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0015]: cannot call non-const fn `<i128 as std::convert::From<u64>>::from` in constants
 --> src/main.rs:7:16
  |
7 |     B(usize) = i128::from(u64::MAX) + 1,
  |                ^^^^^^^^^^^^^^^^^^^^
  |
  = note: calls in constants are limited to constant functions, tuple structs and tuple variants

error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0015`.
Original diagnostics will follow.

warning: the feature `repr128` is incomplete and may not be safe to use and/or cause compiler crashes
 --> src/main.rs:1:12
  |
1 | #![feature(repr128)]
  |            ^^^^^^^
  |
  = note: see issue #56071 <https://github.com/rust-lang/rust/issues/56071> for more information
  = note: `#[warn(incomplete_features)]` on by default

warning: casting `u64` to `i128` may become silently lossy if you later change the type
 --> src/main.rs:7:16
  |
7 |     B(usize) = u64::MAX as i128 + 1,
  |                ^^^^^^^^^^^^^^^^ help: try: `i128::from(u64::MAX)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_lossless
  = note: requested on the command line with `-W clippy::cast-lossless`

warning: `a` (bin "a") generated 2 warnings (1 duplicate)
warning: `a` (bin "a" test) generated 2 warnings (1 duplicate) (run `cargo fix --bin "a" --tests` to apply 1 suggestion)
```

### Version

```text
rustc 1.67.0-nightly (b833ad56f 2022-11-18)
binary: rustc
commit-hash: b833ad56f46a0bbe0e8729512812a161e7dae28a
commit-date: 2022-11-18
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
